perun-core/src/main/java/cz/metacentrum/perun/core/blImpl/AttributesManagerBlImpl.java

```diff
@@ -4508,10 +4517,80 @@ public Attribute mergeAttributeValue(PerunSession sess, User user, Attribute att
 					 return attribute;
 	}
 
+	public Attribute mergeAttributeValue(PerunSession sess, Member member, Attribute attribute) throws InternalErrorException, WrongAttributeValueException,
+			WrongReferenceAttributeValueException, WrongAttributeAssignmentException {
+		// Check type ArrayList
+		if (attribute.getType().equals(ArrayList.class.getName()) || attribute.getType().equals(BeansUtils.largeArrayListClassName)) {
+			Attribute storedAttribute = null;
+			try {
+				// Get current values
+				storedAttribute = getPerunBl().getAttributesManagerBl().getAttribute(sess, member, attribute.getName());
+			} catch (AttributeNotExistsException e) {
+				throw new ConsistencyErrorException(e);
+			}
+
+			//if attribute to merge has null value, return the stored attribute
+			convertEmptyAttrValueToNull(attribute);
+			if(attribute.getValue() == null) return storedAttribute;
+
+			if (storedAttribute != null) {
+				ArrayList<String> updatedList = (ArrayList<String>) storedAttribute.getValue();
+				// If there were someting then find values which haven't been already stored
+				if (updatedList != null) {
+					for (String value : ((ArrayList<String>) attribute.getValue())) {
+						if (!updatedList.contains(value)) {
+							updatedList.add(value);
+						}
+					}
+					attribute.setValue(updatedList);
+				} else {
+					attribute.setValue(attribute.getValue());
+				}
+			}
+
+			getPerunBl().getAttributesManagerBl().setAttribute(sess, member, attribute);
+		} else if (attribute.getType().equals(LinkedHashMap.class.getName())) {
+			Attribute storedAttribute = null;
+			try {
+				// Get current values
+				storedAttribute = getPerunBl().getAttributesManagerBl().getAttribute(sess, member, attribute.getName());
+			} catch (AttributeNotExistsException e) {
+				throw new ConsistencyErrorException(e);
+			}
+
+			//if attribute to merge has null value, return the stored attribute
+			if(attribute.getValue() == null) return storedAttribute;
+
+			// If there were someting then find values which haven't been already stored
+			if (storedAttribute != null) {
```

Comment:
Up to this point the code is exactly same as in the previous if-branch. Consider merging them together and move the branching based on attribute value type to his point.